modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessSemanticAnalysisPhase.java

```diff
@@ -230,4 +235,46 @@ public void decorateWithCastForReturn(
             semanticScope.putDecoration(userExpressionNode, new ExpressionPainlessCast(painlessCast));
         }
     }
+
+    @Override
+    public void visitCallLocal(ECallLocal userCallLocalNode, SemanticScope semanticScope) {
+        if ("$".equals(userCallLocalNode.getMethodName())) {
+            ScriptScope scriptScope = semanticScope.getScriptScope();
+            PainlessMethod thisMethod = scriptScope.getPainlessLookup()
+                .lookupPainlessMethod(scriptScope.getScriptClassInfo().getBaseClass(), false, "field", 1);
+
+            if (thisMethod == null) {
+                throw new IllegalArgumentException("invalid shortcut [$] for [field]; ensure [field] exists in this context");
+            }
+
+            semanticScope.setUsesInstanceMethod();
+            semanticScope.putDecoration(userCallLocalNode, new ThisPainlessMethod(thisMethod));
+
+            scriptScope.markNonDeterministic(thisMethod.annotations.containsKey(NonDeterministicAnnotation.class));
+            Class<?>[] typeParameters = new Class<?>[] { String.class, def.class };
```

Comment:
Can you add a note that we're using `def` here as the type of the default argument due to duck typing?

Reply:
Added a comment.